File: research/experiments/sweep/runs/opt-small-swift-sonnet45/Sources/Optimization/VecOps.swift
import Foundation

/// Pure vector arithmetic for n-dimensional optimization.
/// All operations return new arrays and never mutate inputs.
public enum VecOps {

    /// Dot product of two vectors
    public static func dot(_ a: [Double], _ b: [Double]) -> Double {
        var sum = 0.0
        for i in 0..<a.count {
            sum += a[i] * b[i]
        }
        return sum
    }

    /// Euclidean (L2) norm
    public static func norm(_ v: [Double]) -> Double {
        return sqrt(dot(v, v))
    }

    /// Infinity norm (max absolute value)
    public static func normInf(_ v: [Double]) -> Double {
        var maxAbs = 0.0
        for x in v {
            let absX = abs(x)
            if absX > maxAbs {
                maxAbs = absX
            }
        }
        return maxAbs
    }

    /// Scalar multiplication
    public static func scale(_ v: [Double], _ s: Double) -> [Double] {
        return v.map { $0 * s }
    }

    /// Element-wise addition
    public static func add(_ a: [Double], _ b: [Double]) -> [Double] {
        var result = [Double]()
        result.reserveCapacity(a.count)
        for i in 0..<a.count {
            result.append(a[i] + b[i])
        }
        return result
    }

    /// Element-wise subtraction
    public static func sub(_ a: [Double], _ b: [Double]) -> [Double] {
        var result = [Double]()
        result.reserveCapacity(a.count)
        for i in 0..<a.count {
            result.append(a[i] - b[i])
        }
        return result
    }

    /// Element-wise negation
    public static func negate(_ v: [Double]) -> [Double] {
        return scale(v, -1.0)
    }

    /// Deep copy of vector
    public static func clone(_ v: [Double]) -> [Double] {
        return Array(v)
    }

    /// Create vector of n zeros
    public static func zeros(_ n: Int) -> [Double] {
        return Array(repeating: 0.0, count: n)
    }

    /// Fused operation: a + s*b (avoids intermediate allocation)
    public static func addScaled(_ a: [Double], _ b: [Double], _ s: Double) -> [Double] {
        var result = [Double]()
        result.reserveCapacity(a.count)
        for i in 0..<a.count {
            result.append(a[i] + s * b[i])
        }
        return result
    }
}

File: research/experiments/sweep/runs/opt-medium-python-opus45/src/vec_ops.py
"""
Pure vector arithmetic for n-dimensional optimization.
All operations return new arrays and never mutate inputs.
"""

import math


def dot(a: list[float], b: list[float]) -> float:
    """Dot product of two vectors."""
    return sum(ai * bi for ai, bi in zip(a, b))


def norm(v: list[float]) -> float:
    """Euclidean (L2) norm."""
    return math.sqrt(sum(vi * vi for vi in v))


def norm_inf(v: list[float]) -> float:
    """Infinity norm (max absolute value)."""
    if not v:
        return 0.0
    return max(abs(vi) for vi in v)


def scale(v: list[float], s: float) -> list[float]:
    """Scalar multiplication."""
    return [vi * s for vi in v]


def add(a: list[float], b: list[float]) -> list[float]:
    """Element-wise addition."""
    return [ai + bi for ai, bi in zip(a, b)]


def sub(a: list[float], b: list[float]) -> list[float]:
    """Element-wise subtraction."""
    return [ai - bi for ai, bi in zip(a, b)]


def negate(v: list[float]) -> list[float]:
    """Element-wise negation."""
    return [-vi for vi in v]


def clone(v: list[float]) -> list[float]:
    """Deep copy of a vector."""
    return v[:]


def zeros(n: int) -> list[float]:
    """Vector of n zeros."""
    return [0.0] * n


def add_scaled(a: list[float], b: list[float], s: float) -> list[float]:
    """Compute a + s*b (fused, avoids intermediate allocation)."""
    return [ai + s * bi for ai, bi in zip(a, b)]
